ci/remote-watch.py

```diff
@@ -0,0 +1,247 @@
+            time.sleep(timeout)
+            os.kill(0, signal.SIGKILL)
+    except OSError as ex:
+        if ex.errno not in (errno.EBADF, errno.ESRCH):
+            raise
+        logger.error("Kill error %s: %s", ex.errno, ex.strerror)
+        result = ex.errno
+    return result
+
+
+def yield_poll_schedule():
+    schedule = [0, 5, 5, 10, 20, 40, 40] + [60] * 5 + [120] * 10 + [300, 600]
+    for item in schedule:
+        yield item
+    while True:
+        yield schedule[-1]
+
+
+def detect_spurious_commit(actual, expected, remote):
+    # GitHub sometimes spuriously generates commits multiple times with
+    # different dates but identical contents. See here:
+    # https://github.com/travis-ci/travis-ci/issues/7459#issuecomment-601346831
+    # We need to detect whether this might be the case, and we do so by
+    # comparing the commits' contents ("tree" objects) and their parents.
+    # If these match, we tell the caller to expect the new commit from now on.
+    # Otherwise, we return the same expected commit as before.
+    actual_hash = actual.split(None, 1)[0]
+    expected_hash = expected.split(None, 1)[0]
+    relevant = ["tree", "parent"]  # relevant parts of a commit for comparison
+    if actual != expected:
+        git("fetch", "-q", remote, actual_hash)
+        actual_info = get_commit_metadata(actual_hash)
+        expected_info = get_commit_metadata(expected_hash)
+        a = [pair for pair in actual_info if pair[0] in relevant]
+        b = [pair for pair in expected_info if pair[0] in relevant]
+        if a == b:
+            expected = actual
+    return expected
+
+
+def monitor():
+    ci = get_current_ci() or ""
+    (ref, remote, expected_sha) = git_remote_branch_info()
+    expected_line = "{}\t{}".format(expected_sha, ref)
+    keep_alive = False
+    commit_msg = git("show", "-s", "--format=%B", "HEAD^-")
+    for line in commit_msg.splitlines():
+        parts = line.strip("# ").split(":", 1)
+        (key, val) = parts if len(parts) > 1 else (parts[0], "")
+        if key == "CI_KEEP_ALIVE":
+            ci_names = val.replace(",", " ").lower().split() if val else []
+            if len(ci_names) == 0 or ci.lower() in ci_names:
+                keep_alive = True
+    if keep_alive:
+        logger.info("Not monitoring %s on %s due to keep-alive on: %s",
+                    ref, remote, expected_line)
+        return
+    logger.info("Monitoring %s (%s) for changes in %s: %s", remote,
+                get_remote_url(remote), ref, expected_line)
+    terminate = False
+    for to_wait in yield_poll_schedule():
+        time.sleep(to_wait)
+        status = 0
+        line = None
+        try:
+            # Query the commit on the remote ref (without fetching the commit).
+            line = git("ls-remote", "--exit-code", remote, ref)
+        except subprocess.CalledProcessError as ex:
+            status = ex.returncode
+
+        if status == 2:
+            terminate = True
```

Comment:
terminate is True for all cases that break out of the loop, so this variable is uncessary?

Reply:
Ah there used to be another `break` that we took out, it's redundant now.